### PR TITLE
CONSOLE-3823: Split auth config to its own module

### PR DESCRIPTION
**Pull request by stlaz**
`cmd/main.go` is too big and very hard to navigate through.

Do the first step towards a more reasonable options model.

### CODE REVIEW

File: pkg/server/server.go [hunk 1]
```diff
@@ -417,15 +377,8 @@ func (s *Server) HTTPHandler() http.Handler {
 
 			// TODO remove multicluster
 			authHandlerWithUser(func(user *auth.User, w http.ResponseWriter, r *http.Request) {
-				cluster := serverutils.GetCluster(r)
```
Comment (stlaz) [inline pkg/server/server.go:420]: for the reviewer: I was not sure which chunk exactly is supposed to go away in MC scenarios. LMK if the whole thing can go away.

**Comment by stlaz** [comment]
rebased on top of https://github.com/openshift/console/pull/13244

**Comment by stlaz** [comment]
+ added completion of the authn opetions to finish the opts separation

**Comment by stlaz** [comment]
/retest-required
build error:
```
error An unexpected error occurred: "https://registry.yarnpkg.com/monaco-editor/-/monaco-editor-0.33.0.tgz: unexpected end of file".
```
looks like it works again though

File: cmd/bridge/config/auth/authoptions.go [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+	ClientID     string
+	ClientSecret string
+	CAFilePath   string
+
+	InactivityTimeoutSeconds int
+	LogoutRedirectURL        *url.URL
+}
+
+func NewAuthOptions() *AuthOptions {
+	return &AuthOptions{}
```
Comment (TheRealJon) [inline cmd/bridge/config/auth/authoptions.go:52]: set default values here instead of in 'Complete'?

File: cmd/bridge/main.go [hunk 3]
```diff
@@ -335,20 +300,10 @@ func main() {
...
 
-	var secureCookies bool
-	if baseURL.Scheme == "https" {
-		secureCookies = true
-		klog.Info("cookies are secure!")
-	} else {
-		secureCookies = false
-		klog.Warning("cookies are not secure because base-address is not https!")
-	}
-
```
Comment (TheRealJon) [inline cmd/bridge/main.go:346]: Why is this removed?

File: pkg/flags/validate.go [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+	}
+
+	if ur == nil || ur.String() == "" || ur.Scheme == "" || ur.Host == "" {
+		return nil, NewInvalidFlagError(name, "malformed URL")
+	}
+
+	return ur, nil
+}
+
+func ValidateFlagIs(name string, value string, expectedValues ...string) string {
```
Comment (TheRealJon) [inline pkg/flags/validate.go:58]: This doesn't appear to work as expected. Need to return an error if validation fails.

File: cmd/bridge/main.go [hunk 5]
```diff
@@ -532,127 +494,26 @@ func main() {
...
-	case "disabled":
-		klog.Warning("running with AUTHENTICATION DISABLED!")
-	default:
-		bridge.FlagFatalf("user-auth", "must be one of: oidc, openshift, disabled")
-	}
-
 	switch *fK8sAuth {
 	case "service-account":
-		bridge.ValidateFlagIs("k8s-mode", *fK8sMode, "in-cluster")
+		flags.ValidateFlagIs("k8s-mode", *fK8sMode, "in-cluster")
```
Comment (TheRealJon) [inline cmd/bridge/main.go:499]:
```suggestion
		flags.FatalIfFailed(flags.ValidateFlagIs("k8s-mode", *fK8sMode, "in-cluster"))
```

File: cmd/bridge/main.go [hunk 6]
```diff
@@ -532,127 +494,26 @@ func main() {
...
 		srv.StaticUser = &auth.User{
 			Token: k8sAuthServiceAccountBearerToken,
 		}
 		srv.ServiceAccountToken = k8sAuthServiceAccountBearerToken
 	case "bearer-token":
-		bridge.ValidateFlagNotEmpty("k8s-auth-bearer-token", *fK8sAuthBearerToken)
+		flags.ValidateFlagNotEmpty("k8s-auth-bearer-token", *fK8sAuthBearerToken)
+		flags.FatalIfFailed(err)
```
Comment (TheRealJon) [inline cmd/bridge/main.go:506]:
```suggestion
		flags.FatalIfFailed(flags.ValidateFlagNotEmpty("k8s-auth-bearer-token", *fK8sAuthBearerToken))
```

File: cmd/bridge/main.go [hunk 7]
```diff
@@ -532,127 +494,26 @@ func main() {
...
+
 		srv.StaticUser = &auth.User{
 			Token: *fK8sAuthBearerToken,
 		}
 		srv.ServiceAccountToken = *fK8sAuthBearerToken
 	case "oidc", "openshift":
-		bridge.ValidateFlagIs("user-auth", *fUserAuth, "oidc", "openshift")
+		flags.ValidateFlagIs("user-auth", authOptions.AuthType, "oidc", "openshift")
```
Comment (TheRealJon) [inline cmd/bridge/main.go:513]:
```suggestion
		flags.FatalIfFailed(flags.ValidateFlagIs("user-auth", authOptions.AuthType, "oidc", "openshift"))
```

File: pkg/serverconfig/validate.go [hunk 8]
```diff
@@ -19,7 +19,7 @@ func Validate(fs *flag.FlagSet) error {
 		return err
 	}
 
-	bridge.ValidateFlagIs("user-settings-location", fs.Lookup("user-settings-location").Value.String(), "configmap", "localstorage")
+	flags.ValidateFlagIs("user-settings-location", fs.Lookup("user-settings-location").Value.String(), "configmap", "localstorage")
```
Comment (TheRealJon) [inline pkg/serverconfig/validate.go:22]:
```suggestion
	flags.FatalIfFailed(flags.ValidateFlagIs("user-settings-location", fs.Lookup("user-settings-location").Value.String(), "configmap", "localstorage"))
```

**Review by TheRealJon — CHANGES_REQUESTED** [review]
Just a few things that I think might have been missed. This is definitely an improvement!

File: pkg/flags/flags.go
[the remark this answers is not in the document]
Comment (stlaz) [reply pkg/flags/flags.go:1]: Ah, right, this is from a follow up. I'll remove it here and then have a look at your suggestion there, looks like exactly what I need :+1:

File: cmd/bridge/main.go [hunk 9]
```diff
@@ -143,59 +136,47 @@ func main() {
...
 	if err := serverconfig.Validate(fs); err != nil {
 		fmt.Fprintln(os.Stderr, err.Error())
 		os.Exit(1)
 	}
 
-	baseURL := &url.URL{}
-	if *fBaseAddress != "" {
-		baseURL = bridge.ValidateFlagIsURL("base-address", *fBaseAddress)
-	}
+	authOptions.ApplyConfig(&cfg.Auth)
```
[the remark this answers is not in the document]
Comment (stlaz) [reply cmd/bridge/main.go:150]: good catch!

File: cmd/bridge/config/auth/authoptions.go (hunk 2 above)
Comment (stlaz) [reply cmd/bridge/config/auth/authoptions.go:52]: We can only do that after `ApplyConfig()` is called

File: cmd/bridge/main.go (hunk 3 above)
Comment (stlaz) [reply cmd/bridge/main.go:346]: this is `useSecureCookies = baseURL.Scheme == "https"` in `getAuthenticator()`

File: pkg/flags/validate.go (hunk 4 above)
Comment (stlaz) [reply pkg/flags/validate.go:58]: :man_facepalming: another great catch

**Review by TheRealJon — APPROVED** [review]
/lgtm

**Comment by TheRealJon** [comment]
QE Approver
/assign @yanpzhan

Docs Approver
/assign @opayne1

PX Approver
/assign @RickJWagner

**Comment by yanpzhan** [comment]
@stlaz @TheRealJon hi, I'm trying to test the pr, but not sure about the test coverage? if idp related regression test is enough? Could you give me some guide about checkpoints?

**Comment by jhadvig** [comment]
/retest 

```
  Running:  cluster-settings/alertmanager/alertmanager.cy.ts                              (35 of 44)
[1024:1113/112155.965205:ERROR:zygote_host_impl_linux.cc(263)] Failed to adjust OOM score of renderer with pid 9664: Permission denied (13)
  Alertmanager
This socket has been ended by the other party
Error: This socket has been ended by the other party
    at TLSSocket.writeAfterFIN [as write] (node:net:487:14)
    at ClientRequest.<anonymous> (<embedded>:4536:1112)
    at ClientRequest.emit (node:events:527:28)
    at HTTPParser.parserOnIncomingClient [as onIncoming] (node:_http_client:631:27)
    at HTTPParser.parserOnHeadersComplete (node:_http_common:128:17)
    at TLSSocket.socketOnData (node:_http_client:494:22)
    at TLSSocket.emit (node:events:527:28)
    at addChunk (node:internal/streams/readable:315:12)
    at readableAddChunk (node:internal/streams/readable:289:9)
    at Readable.push (node:internal/streams/readable:228:10)
    at TLSWrap.onStreamRead (node:internal/stream_base_commons:190:23)
error Command failed with exit code 1.
```

**Comment by yanpzhan** [comment]
Thanks. Regression test about idp tests passed on the cluster launched against the pr.
/label qe-approved
